### DISCUSSION TITLE
I'm making a small game with SDL and Lua

### DISCUSSION THREAD

**Discussion by eliasdaler** [eliasdaler/edw.is#4] [General]
Okay, trying something new. It'll be easy for me to write "throw away" posts here and not pollute my main website.

Then I might compile it into a proper article, but it will take some time.  You can subscribe to this discussion to get new updates! Neat.

---

My previous Lua/C++ games had something like this:

* 40% of code is C++
* 60% Lua

Now I want to make something closer to LOVE2D - it'll be a wrapper around SDL/OpenGL and *most* of the code will be written in Lua to not deal with C++ nonsense.

For example, in my previous games ECS was implemented on C++ side, so you had to do a lot of bindings to create new components/systems in Lua. Now I want to implement the whole ECS thing in Lua:

1. Every component/system etc. will be in Lua
2. I won't have to deal with template programming which made C++ ECS system very difficult to maintain.

I think I'll open-source it,  if something good comes out of it. 

At least I'll document my findings in a "blog" format. 

Please post comments to each post as a reply, not as a new post in this thread. :)

---

P.S. Making stuff from "almost" scratch is supposed to be easy. But it isn't, and we need to think why it is so and how this can be changed without using full-blown game engines.

**Comment by eliasdaler** [comment]
# Day 1-2. Window, sprite moving around

Starting... The current plan is to port [Ebiten Breakout](https://github.com/eliasdaler/ebiten_breakout) and see how much code it'll take. 

Then I'll make something new.

# Dependencies

third_party dir looks like this:

```sh
third_party/
├── imgui 
├── lua
├── SDL
├── SDL_image
├── sol2
└── CMakeLists.txt
```

Things I need to add:

* glm - very nice matrices, vectors, quaternions 
* something for audio (might use SDL_audio at first, but want to play tracker music eventually)
* assimp for loading 3D models
* spdlog for logging

# Current progress

The game looks like this:

![image](https://user-images.githubusercontent.com/1285136/219685044-5b74fb3d-7637-4087-8545-714462364c25.png)

Lua script looks like this:

```lua
local sprite
local x = 100
local y = 100
local speed = 600

function init(texture)
    sprite = Sprite.new(texture)
    sprite:setTextureRect(0, 0, 16, 16)
end

function update(dt)
    local dx = 0
    local dy = 0

    if isKeyPressed(UP) then
        dy = dy - 1
    end

    if isKeyPressed(DOWN) then
        dy = dy + 1
    end

    if isKeyPressed(LEFT) then
        dx = dx - 1
    end
    if isKeyPressed(RIGHT) then
        dx = dx + 1
    end

    x = x + speed * dx * dt
    y = y + speed * dy * dt

    sprite:setPosition(x, y)
end

function draw()
    drawSprite(sprite)
end
```

So, I got the basic "sprite" class and can check if a keyboard key was pressed or not. Cool!

# Graphics

I'm using SDL_Renderer for rendering for now. I want shader and 3D graphics support, so I'm thinking of trying out something from this set:

* bgfx
* sokol
* magnum
* DiligentEngine
* OpenGL 3 (I have quite of few stuff implemented from my previous engine: models, skeletal animations, shaders, sprites, billboards, etc.)

I have great hopes for [this](https://github.com/libsdl-org/SDL_shader_tools) - hope I'll be able to use it in the future to not think about OpenGL/Vulkan/Metal etc. ever again. 

I wish software rendering was a viable option since I don't want to make a AAA game - the PCs of today are powerful enough to handle it, but I'll have to do quite a lot of things that GPU does for you manually, so it might be worse than using OpenGL.

# Stats

```
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
C++                              4             59             23            202
C/C++ Header                     3             20              0             50
Lua                              1              8              0             30
CMake                            1              4              2             16
-------------------------------------------------------------------------------
SUM:                             9             91             25            298
-------------------------------------------------------------------------------
```

**Reply by a327ex** [reply]
What a coincidence, I literally just started something similar yesterday. SDL and Lua too, but I'm planning on sticking to software rendering alone with no OpenGL since my current Lua engine (built on top of LÖVE) has by accident converged into an API that I think will work really well with just software rendering. Plus I feel like I often try to reach for the ability to set pixels individually, so I'm wondering if doing it like this is even a feasible option, I never tried it so I don't know how slow it is.

In any case, good luck with your explorations!

**Reply by eliasdaler** [reply]
Interesting. Is there any place where I can see your progress? Or is it private for now?

I think that software renderers are really underappreciated. You can do almost anything with one function:

```lua
SetPixel(x, y, r, g, b)
```

That's it. And with Lua and its portability, you can potentially port it to any platform - you just need a way to run Lua's VM and ability to draw a pixel in a given point. That's the **ultimate** minimalism and interesting to think about.

**Reply by a327ex** [reply]
I'll keep it private for now but I'll eventually make it public. And yea, the simplicity of it is really attractive to me both from an implementation perspective but also creatively. There are a bunch of things I want to do that don't feel very intuitive to me with the standard GPU abstractions. Hopefully it works well. I'll keep reading your updates too, good luck!

**Reply by jessechounard** [reply]
This is cool! I'm working on a new port of a game to iOS after being away from gamedev for years. Discovering that OpenGL and OpenAL are deprecated there was a big shock, so I'm also using the built in SDL renderer. Since it has built-in batching, when the shader stuff comes out, it'll have most of what I need. If you end up switching to bgfx, I'd definitely love to hear your thoughts on it.

**Reply by eliasdaler** [reply]
@a327ex, thank you! If you stumble upon interesting "software rendering" resources, I'd be interested in them. I feel like most of them focus on ray-tracing, and that's not really what I need.

@jessechounard, I will use SDL_renderer for now to make a "basic" game and then I'll try to replace it with something complex once I start trying to make 3D stuff. :)

**Comment by eliasdaler** [comment]
# Day 3-4 Working game!

![](https://i.ibb.co/9vyrJyh/breakout.gif)

[This game](https://github.com/eliasdaler/ebiten_breakout) is almost ported. Stuff to port:

* Viewport scaling (the GIF was recorded at 256x240 and upscaled 3x)
* Music and audio
* States (menu, game over and win screens)

I'm reusing "Go themed" sprites for now, but will redraw them to be "Lua, C++ themed" soon. The ball will be Lua's symbol, of course. But what should the paddle and blocks look like? Hmm...

---

Stats:

```
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Lua                              1             54             20            253
C++                              4             55             14            207
C/C++ Header                     3             19              0             47
CMake                            1              4              2             16
-------------------------------------------------------------------------------
SUM:                             9            132             36            523
-------------------------------------------------------------------------------
```
---

OK, it was easy. I'm porting this game to be sure that I can make something "complete" and not get stuck worrying about the small stuff.

As your can see, I now have more Lua code than C++ code. Isn't it great? For now, the C++ parts that Lua code calls is pretty minimal:

* ```isKeyPressed``` - a wrapper around `SDL_GetKeyboardState`
* ```drawSprite``` a wrapper around small `Sprite` class, which contains `SDL_Texture*` and calls `SDL_RenderCopy` internally
* ```loadTexture``` loads `SDL_Texture*`, uses SDL_image
... and that's it.

I think I might simplify `Sprite` class even further how Love2D does and have interface like this:

```lua
local s = sprite.New("some_texture.png")
s:setTextureRect(32, 32, 16, 16)
```
I know about techniques like batching textures into one mega-texture (like Ebitengine does), but for now I don't want to go that route and focus on other things - I can optimize it later if I see the need to do so. 

And if I decide to go with software rendering for fun, I can even replace `drawSprite` with a function which calls `SetScreenPixel` a bunch of times. 

I think that software rendering can work out pretty well for the 2D games I want to make:

* I'm a pixel purist, so I want to make games in 256x224 resolution or similar (like SNES does) 
* No sprite scaling
* No sprites of different sizes
* No sprite rotations / pixel interpolation
* No complex alpha multiply/add (for now)
* No shaders

The only shader I want to have is a CRT shader - but I can do it in OpenGL *after* I've rendered everything I want to be present on the screen to a texture. So, full screen effects will be optional and won't make "the base* rendering engine harder than it needs to be.

If I decide to do 3D (and I might!), I'll totally reuse my OpenGL rendering code for now. I've found [g3d](https://github.com/groverburger/g3d) to be inspiring and might steal some API ideas from it.

I also learned that SDL's Switch port works even if you write stuff with OpenGl ([source](https://twitter.com/icculus/status/981733067369996288)), so that's pretty awesome.

---

But for the present time, I want to focus on the **fun** parts - tile maps, NPCs and dialogues, cutscenes, etc.

So, gameplay/stuff that every game needs first, graphics and niceties second. :)

**Comment by eliasdaler** [comment]
# Off-topic. Writing code in Lua vs Go

For now, I'm suffering a bit, to be honest. Writing `local x = 5` is not as pleasant as writing `x := 5`.

This stuff is also frustrating:

```lua
-- script.lua
local function f()
    g(5) -- error: g is nil

local function g(x)
   print(x)
end
```

Lua is a one-pass compiler, so if it doesn't see "g" at first, it assumes that it's a global variable. To fix it, you need to write this:

```lua
-- script.lua
local g
local function f()
    g(5) -- error: g is nil

g = function(x)
   print(x)
end
```
... not as good as Go, where you can define functions everywhere you want... Plus, it breaks the pattern which I've grown to love:

```go
func f() {
   f1()
   f2()
   f3()
}

func f1() { ... }
func f2() { ... }
func f3() { ... }
```

So, you read the source from top to bottom. First, you read the higher level functions, then you scroll down and read more "low level" functions. Lua makes it a bit harder to write. Of course, you can "fix" it by using "OOP-like" approach, e.g.:

```lua
local Game = {}

function Game:init()
    self:spawnObjects()
end

function Game:spawnObjects()
end
```

... and it sort of works. But it makes writing free functions a bit harder. Oh well.

---

Then, you have other problems of dynamic languages

* Typos everywhere ("was it `x.w` or `x.width`?")
* If you refactor, you need to be very careful. The compiler won't save you
* Tables are not copied by default, which makes working with tables tricky if you want them to simulate "classes", e.g.

```lua
local pos = { x = 20, y = 30 }
local newPos = pos
newPos.x = 30 -- oops, "pos" changed too!
```
With Go, it wasn't like this - every struct is allocated on stack by default and so when you do `x = y`, you **copy** stuff, which is more pleasant.

---

I think I'll rant more in the coming days, but I hope that I forget the "Go" stuff (since I won't be writing much of it anymore) and rewire my brain to Lua. 

It's not all gloom, of course. Not writing types and being able to easily "extend" anything by just doing `someTable.newField = "blah"` is nice. Simplicity is nice. 

I'm also reading "Programming in Lua" again and it's so great. No filler, just the stuff you need to know.

**Comment by eliasdaler** [comment]
# Day 5-8 First game - done

And here it is!
The small, but surprisingly non-trivial breakout game. I'm releasing the source code to public domain.

Here it is:

https://github.com/eliasdaler/sdl_breakout

![image](https://i.ibb.co/1LdJktH/gameplay-breakout.gif)

This is a port of my old test game, [Ebiten Breakout](https://github.com/eliasdaler/ebiten_breakout).

It might be small, but I'm sure it might be useful to some people. :)

Final stats (excluding CMake)

```
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Lua                              7            127             24            560
C++                              4            109             49            425
C/C++ Header                     4             20              0             68
-------------------------------------------------------------------------------
SUM:                            15            256             73           1053
-------------------------------------------------------------------------------
```

The similar game which I wrote in Ebitengine and Go is 834 lines of code. Considering that Ebitengine is much higher level than SDL, that's not too bad.

A lot of interesting things have happened since I last posted.

## Engine / game separation

The game is now fully separated from the engine (I call it "edge"). The game is fully written in Lua.

The engine's C++ API is pretty simple. Here it is:

```lua
-- graphics
edge.loadTexture(path) -- loads a texture
edge.loadFont(path, size) -- loads a TTF font

drawableString = edge.newDrawableString(font, text) -- creates a text string which can be drawn
drawableString:draw(x,y,r,g,b) -- draws a string with given color

sprite = edge.newSprite(texture) -- creates a sprite
sprite:setTextureRect(x,y,w,h) -- sets a part of texture to be displayed
sprite:getTextureRect() -- returns a texture rect
sprite:draw(x,y)

-- sound
sound = edge.loadSound(path) -- loads a sound
sound:play() -- plays a sound

music = edge.loadMusic(path) -- loads music (in .ogg)
music:play() -- plays music
music:stop() -- stop music

-- input
edge.isKeyPressed(key) -- checks is keyboard key is pressed
```

Here are some thoughts on this API design.

1. You'll see (x,y) and (r,g,b) being thrown around. While this is great for simplicity, I feel like at some point I'll make proper vector and color classes and accept them on C++ side - they're much easier to handle and less error-prone
2. I use `edge.newSprite` instead of `edge.Sprite.new(...)` for two reasons:
    * In sol2, you create objects with `.new(..)`, but in middleclass (the lib I use for OOP), you need to use `:new()` - this is inconsistent and error-prone - so it's better to make C++ constructors/factory functions feel "different"
    * It looks like Love2D's `love.graphics.newX` and feels nice :)

Speaking of Love2D... I wonder why they chose to separate things by modules like `love.graphics.newSprite`, instead of just doing `love.newSprite` - the former makes Lua do an additional lookup, but maybe it looks nicer?

The engine also has some parts writte in in Lua. They're stored in "edge/src/Lua" - I feel like it's nice to write "middle level" stuff like this. This is where FSMs, event manager, ECS and other things will go in the future.

## Launcher

Theoretically, each game doesn't need to have a specific launcher. It can just execute "scripts/game.lua" and that's where all the game code will start. However, I found it difficult to make a CMake target which will work like this and properly link to stuff. I will work on it later.

Love2D also has this interesting feature, which allows you to "zip" all your resources and scripts, and then just "cat" them with a Love2D's binary to get a runnable executable. This is great and simple, I might look into it.

Being able to make a "single file" binary for your entire game sounds pretty neat (not that great for modding, though!)

## CMake

You will notice that this game has tons of CMake code. This is caused by the following:
* Building all dependencies myself
* Doing Lua scripting (you need to symlink/copy directories to make Lua find your scripts)
* Making assets folder easy to update and track with git by symlinking to it in build directory

You will notice that I use symlinks during build step. I've found this much better than copying scripts/assets to build directory, because you'll never accidentally modify a script inside the build dir (instead of source dir, as you probably want).

I still didn't manage dynamic linking to work on Windows. You also need to copy DLLs to install/build dir there. I'll work on it later.

## Asserts are great

Another great thing I've figured out is this pattern:

```lua
self.font = assert(edge.loadFont("res/font.ttf", 8))
```

`loadFont` returns two values, a font and an error message if loading fails.
What's great about `assert` is that it automatically creates an error with a returned message if the first return of `loadFont` is `nil`.

This means, I don't need to write code like this:
```lua
local msg
self.font, msg = edge.loadFont("res/font.ttf", 8)
if not self.font then
    error(msg)
end
```

Doing this in one of code is beatiful. And I can also catch this error on Lua side if I don't want it to "escape" to C++.
And if it's uncaught, I simply print it to console and exit:

```
scripts/game.lua:71: Couldn't open res/fonty.ttf
stack traceback:
	[C]: in function 'base.assert'
	scripts/game.lua:71: in method 'initialize'
	...work/games/build/games/breakout/edge/lib/middleclass.lua:147: in function <...work/games/build/games/breakout/edge/lib/middleclass.lua:144>
	(...tail calls...)
```

As you can see, the error message also shows me the line on which the error has occured. Nice!


## C?

Theoretically, all the C++ code could be ported to C. I'll need to get rid of sol2 binding code and write *a lot* of Lua C API code. I don't want to do it, so I'll use it for now.

But it's interesting how the breakout itself can now even be ported to Love2D's code without rewriting much of it (you only need to write "glue" code)

## Future work

The next thing will be much bigger. I want to make a small top-down adventure game with tiles, NPCs, dialogue etc. I'll also make a small level editor. This time in Lua (I had a 3k+ LoC of C++ code for level editor in my previous engine).

I'll also stop using `SDL_Renderer`. While it's great, I want shaders and maybe experiment with 3D a little bit. This would be a bit tricky to expose to Lua, but I hope to arrive with a simple API which will hide all the gory details of how things *really* work.

See you later!

**Comment by eliasdaler** [comment]
# Next game: small tile-based top-down game + level editor

Sorry for the lack of updates recently.

I recently ported my graphics code from SDL_Renderer to OpenGL. I used my old code from my previous games (from 2019-2020) and turns out it was *pretty* good, because I was able to quickly adapt it. Mainly I just used it to do simple sprites/quad buffers.

Here's what I have now (notice the nice CRT shader, can't do that with SDL_Renderer!): 
![image](https://user-images.githubusercontent.com/1285136/222934287-5701dabd-d231-464e-9574-6f75d1c6a501.png)

Watch windows (wow, you can even embed videos here!)

https://user-images.githubusercontent.com/1285136/222934650-ee65df70-10f7-420e-9553-fa8ce5bb29d1.mp4

## Things I learned about Lua recently.

1. Putting things into modules like this is nice:

```lua
local M = {}
function M:doSomething()
  ...
end

function M:doAnother()
  M.doSomething()
end

return M
```

This solves the "locals" problem I was talking about previously.

2. Turns out Lua 5.4's <const> is kinda pointless. It only works for local variables. What a shame.
3. This debugger exists: https://github.com/slembcke/debugger.lua. And it's awesome. You just require a module, drop `dbg()` call somewhere and you get GDB-like experience in your Lua code. Amazing.
4. Lua is much faster to write than Go, but you need to be more attentive and careful (typos, typos eveywhere)

## What I'm doing now

Mostly boring stuff. Need to rewrite other "basic" things - level editor stuff, proper input management (*screaming internally*), basic collision detection, etc.-etc.

I might go a bit quite for a while since it's mostly nothing unusual or ground breaking. If you want to hear about some of this, though, let me know! :)

**Reply by lethal-guitar** [reply]
> proper input management (screaming internally)

What would that involve? Why does it make you scream? 😄

**Reply by eliasdaler** [reply]
I'll need to wrap keyboard/gamepad events into "fake gamepad events" so that I can allow remapping any key to a game "action"/"virtual gamepad interaction", e.g. pressing "W" on a keyboard might map to "Left stick Y axis is now -1".
Also need to handle gamepad connect/disconnect, deadzones, etc.

It's not really difficult to do, just tedious and involves a lot of code I wrote two times already. (though I want to do it in Lua now, so I'll need to do it once again)

**Comment by eliasdaler** [comment]
As you can tell, the experiment is over and I've been making a bigger 3D game in C++ for a few months now:

https://github.com/eliasdaler/edw.is/discussions/6
